### PR TITLE
Add time variables for snippets

### LINKED ISSUE #41631: Feature request: expose date information as available variable(s) when creating snippets.
Edit: Sorry if this is a terrible issue...

The current list of variables available when writing a snippet can be found [here](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables).

Existing variables expose great contextual information - such as current file, directory, line number, word, etc. Development happens so fast that information can become outdated, so knowing when some custom snippets were inserted could be useful. For example, when creating files or adding comments.

My current use case is a snippet that inserts copyright headers.

The key binding `ctrl+cmd_c` inserts the header:
```ts
/**
 * MyClass.ts
 * MyProjectName
 *
 * Created by David Hunt.
 * Copyright © Our Company Name, LLC. All rights reserved.
 */
```

It would help to include the date:
```ts
/**
 ...
 * Created by David Hunt on 1/15/18.
 ...
```
and could also apply to updates or comments:
```ts
/**
 * See stackoverflow issue: someurl.
 * Added 1/15/18.
 */
```

Doing so could mean a simple variable like `TM_CURRENT_DATE`, but that may not work for non-US users who's date format is different i/e _not_ month/day/year. Perhaps exposing the variables separately such as `TM_CURRENT_YEAR`, etc or localization can help?

I'd like to know if there is any interest in this. I know an extension could do this but it seems like overkill, but that could depend on potential use cases.

Thanks for any feedback.

### PR DESCRIPTION
**Pull request by usernamehw**
Fixes #41631 

@jrieken

### CODE REVIEW

File: src/vs/editor/contrib/snippet/snippetVariables.ts [hunk 1]
```diff
@@ -170,3 +177,29 @@ export class ClipboardBasedVariableResolver implements VariableResolver {
 		}
 	}
 }
+
+export class TimeBasedVariableResolver implements VariableResolver {
+
+	resolve(variable: Variable): string {
+		const { name } = variable;
+		const zeroPad = (n: string): string => n.length < 2 ? `0${n}` : n;
```
Comment (jrieken) [inline src/vs/editor/contrib/snippet/snippetVariables.ts]: better use the existing util from base/common/strings#pad

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/snippet/snippetVariables.ts (hunk 1 above)
Comment (usernamehw) [reply src/vs/editor/contrib/snippet/snippetVariables.ts]: Didn't know about it

**Review by jrieken — APPROVED** [review]

**Comment by jrieken** [comment]
Would you might adding a test? At least for manifesting the names of the those variables? Maybe have your code so that the date to be used is an optional ctor-argument. Thanks already.

**Comment by usernamehw** [comment]
Travis-CI build stuck > 7hours.

**Comment by jrieken** [comment]
lgtm. Thanks

**Comment by taivu** [comment]
has this been released into a build yet? do I use it like `${CURRENT_YEAR}`?

**Comment by usernamehw** [comment]
@taivu currently it's in [Insiders](https://code.visualstudio.com/insiders/) build

**Pull request merged**